### ISSUE TITLE
Decoded Subject does not match Yahoo Mail's interpretation

### ISSUE THREAD

**Issue by ekalchev** [jstedfast/MimeKit#449]
Subject string is not parsed correctly using MimeMessage.Load

Version of MimeKit
MimeKit Commit: 8cb80e86f9dbcf00e10da410eb782d7e97ca204d [8cb80e8]

This is reproducible using those mail messages
http://s000.tinyupload.com/index.php?file_id=37381762003408438907

![image](https://user-images.githubusercontent.com/10009149/49142071-a8107300-f300-11e8-98c1-83794db28b57.png)

**Comment by jstedfast** [comment]
Actually, I'm not sure what the "correct" thing to do *is*...

Are you thinking that just because the resulting subject looks like a bunch of gobbeldygook that it must be wrong?

Here's what Thunderbird sees (just for comparison):

<img width="743" alt="mimekit-issue449-thunderbird" src="https://user-images.githubusercontent.com/338984/49155068-ffeeb000-f2e7-11e8-91d2-746d5bf50fba.png">

Here's the raw Subject header of the first message:

```
Subject: =?UTF-8?B??==?UTF-32LE?B?=473649888336573TgDpCYhqDaSbuYK?==?UTF-8?B??==?UTF-8?Q?=54?==?UTF-8?B??==?UTF-32LE?B?=5Q0I9DD3J2G26IY7S3t3?==?UTF-8?B??==?UTF-8?Q?=69?==?UTF-8?B??==?UTF-32LE?B?=763R2QOZD8P5a1a4k83L2h?==?UTF-8?B??==?UTF-8?Q?=72?==?UTF-8?B??==?UTF-32LE?B?=6SPPDNO1H9099u1jmBZ61Lk0bB?==?UTF-8?B??==?UTF-8?Q?=65?==?UTF-8?B??==?UTF-32LE?B?=652RIDX86O32Ztxnu86Ee9L0?==?UTF-8?B??==?UTF-8?Q?=64?==?UTF-8?B??==?UTF-32LE?B?=6537695242908316FjaPrcQKGYRKSJuU?==?UTF-8?B??==?UTF-8?Q?=20?==?UTF-8?B??==?UTF-32LE?B?=1EMCE14W83E0FJ38d85KM9Bzq2Fqee?==?UTF-8?B??==?UTF-8?Q?=6F?==?UTF-8?B??==?UTF-32LE?B?=U2R38AG5B91HM0j82WBWW3R14817?==?UTF-8?B??==?UTF-8?Q?=66?==?UTF-8?B??==?UTF-32LE?B?=9D9VUA69B6U107GUE454pjofnYH5Azfl?==?UTF-8?B??==?UTF-8?Q?=20?==?UTF-8?B??==?UTF-32LE?B?=7KUK75P4J4K3EP09Z1y6j6BGqO0cm9uOGbUu?==?UTF-8?B??==?UTF-8?Q?=65?==?UTF-8?B??==?UTF-32LE?B?=M9CV8K2A6K41F20VGFVD9I9s3H21YWI2ZD?==?UTF-8?B??==?UTF-8?Q?=78?==?UTF-8?B??==?UTF-32LE?B?=YIPWYGJQ1RQI2W1QPL1Shm4MjBu3kw4Y5C8yrZ?==?UTF-8?B??==?UTF-8?Q?=70?==?UTF-8?B??==?UTF-32LE?B?=24001486469747446eiqgsHalColCizvbi?==?UTF-8?B??==?UTF-8?Q?=65?==?UTF-8?B??==?UTF-32LE?B?=3U7G3M2EXLLN41TYH1MJBNBFBnV1Q4ciqzQuZw6S?==?UTF-8?B??==?UTF-8?Q?=6E?==?UTF-8?B??==?UTF-32LE?B?=UETJE6ZSX092207RQCXM35kzRI0bf4s25FoZ2t8CQ7?==?UTF-8?B??==?UTF-8?Q?=73?==?UTF-8?B??==?UTF-32LE?B?=0RBQ3RY3IUNOF368M543A3lj2M8RFgGPNMO8tKPzG6jW?==?UTF-8?B??==?UTF-8?Q?=69?==?UTF-8?B??==?UTF-32LE?B?=4I31467T718F28M8XJMW2MIrE4woqd45EzAS8Dd4N5usTj?==?UTF-8?B??==?UTF-8?Q?=76?==?UTF-8?B??==?UTF-32LE?B?=FUY65C8N97P1XMGWO2C62ZLOF9hpw3FWz3kB3aTDiUy1kfLt?==?UTF-8?B??==?UTF-8?Q?=65?==?UTF-8?B??==?UTF-32LE?B?=Y200WZ4I?==?UTF-8?B??==?UTF-8?Q?=20?==?UTF-8?B??==?UTF-32LE?B?=56L8SZLUJ?==?UTF-8?B??==?UTF-8?Q?=4D?==?UTF-8?B??==?UTF-32LE?B?=5Q0I9DD3J2?==?UTF-8?B??==?UTF-8?Q?=65?==?UTF-8?B??==?UTF-32LE?B?=763R2QOZD8P?==?UTF-8?B??==?UTF-8?Q?=64?==?UTF-8?B??==?UTF-32LE?B?=652RIDX86O32?==?UTF-8?B??==?UTF-8?Q?=69?==?UTF-8?B??==?UTF-32LE?B?=4TMom1N?==?UTF-8?B??==?UTF-8?Q?=63?==?UTF-8?B??==?UTF-32LE?B?=PntIRxi0?==?UTF-8?B??==?UTF-8?Q?=61?==?UTF-8?B??==?UTF-32LE?B?=jpHB8ao43?==?UTF-8?B??==?UTF-8?Q?=72?==?UTF-8?B??==?UTF-32LE?B?=G26IY7S3t3?==?UTF-8?B??==?UTF-8?Q?=65?==?UTF-8?B??==?UTF-32LE?B?=5a1a4k83L2h?==?UTF-8?B??==?UTF-8?Q?=20?==?UTF-8?B??==?UTF-32LE?B?=Ztxnu86Ee9L0?==?UTF-8?B??==?UTF-8?Q?=70?==?UTF-8?B??==?UTF-32LE?B?=u1jmBZ61Lk0bB?==?UTF-8?B??==?UTF-8?Q?=72?==?UTF-8?B??==?UTF-32LE?B?=j82WBWW3R14817?==?UTF-8?B??==?UTF-8?Q?=65?==?UTF-8?B??==?UTF-32LE?B?=8d85KM9Bzq2Fqee?==?UTF-8?B??==?UTF-8?Q?=6D?==?UTF-8?B??==?UTF-32LE?B?=E454pjofnYH5Azfl?==?UTF-8?B??==?UTF-8?Q?=69?==?UTF-8?B??==?UTF-32LE?B?=FVD9I9s3H21YWI2ZD?==?UTF-8?B??==?UTF-8?Q?=75?==?UTF-8?B??==?UTF-32LE?B?=y6j6BGqO0cm9uOGbUu?==?UTF-8?B??==?UTF-8?Q?=6D?==?UTF-8?B??==?UTF-32LE?B?=Shm4MjBu3kw4Y5C8yrZ?==?UTF-8?B??==?UTF-8?Q?=73?==?UTF-8?B??==?UTF-32LE?B?=BNBFBnV1Q4ciqzQuZw6S?==?UTF-8?B??==?UTF-8?Q?=3F?==?UTF-8?B??==?UTF-32LE?B?=5kzRI0bf4s25FoZ2t8CQ7?==?UTF-8?B??=
```

Here's the Subject header of the second message:

```
Subject: =?ASCII?B?==F36QBe7gQn?==?UTF-8?B?Rw==?==?ASCII?B?==9HuDy8u?==?UTF-8?B?ZQ==?==?ASCII?B?==aON3qWkkk0Q?==?UTF-8?B?dA==?==?ASCII?B?==K4DbV5Vdy?==?UTF-8?B?IA==?==?ASCII?B?==S5zi1ck5HH?==?UTF-8?B?QQ==?==?ASCII?B?==9ixYeTk0z2YXU7n5?==?UTF-8?B?IA==?==?ASCII?B?==JuVmI94EUJV1g?==?UTF-8?B?JA==?==?ASCII?B?==w2wyGyUM1Ysd?==?UTF-8?B?MQ==?==?ASCII?B?==uB1VfcPIrLZC?==?UTF-8?B?LA==?==?ASCII?B?==1BL6B84A6?==?UTF-8?B?MA==?==?ASCII?B?==RZCEgUXJ?==?UTF-8?B?MA==?==?ASCII?B?==GsZqUpJ?==?UTF-8?B?MA==?==?ASCII?B?==aON3qWkkk0Q?==?UTF-8?B?LA==?==?ASCII?B?==3KEXk2i89cu?==?UTF-8?B?MA==?==?ASCII?B?==SQNYRZtR?==?UTF-8?B?MA==?==?ASCII?B?==A3C8g36xc?==?UTF-8?B?MA==?==?ASCII?B?==K6A7Wm4CHGZK020?==?UTF-8?B?IA==?==?ASCII?B?==Xj84g?==?UTF-8?B?TA==?==?ASCII?B?==tBByrB?==?UTF-8?B?aQ==?==?ASCII?B?==9s1nYJCfaCQS?==?UTF-8?B?Zg==?==?ASCII?B?==45NyojvF90L?==?UTF-8?B?ZQ==?==?ASCII?B?==Iie1j48Ej60NW?==?UTF-8?B?IA==?==?ASCII?B?==KJu9J?==?UTF-8?B?SQ==?==?ASCII?B?==B2Gf6UB7?==?UTF-8?B?bg==?==?ASCII?B?==w00D5lt0dXL?==?UTF-8?B?cw==?==?ASCII?B?==bxjV8XUW?==?UTF-8?B?dQ==?==?ASCII?B?==88uEdeT?==?UTF-8?B?cg==?==?ASCII?B?==2xpgTDG?==?UTF-8?B?YQ==?==?ASCII?B?==xs7S5m3CxUVxyF03?==?UTF-8?B?bg==?==?ASCII?B?==6UsNk96fl?==?UTF-8?B?Yw==?==?ASCII?B?==ZK2Y6t4fxgFcHYD3o?==?UTF-8?B?ZQ==?==?ASCII?B?==Kscws011l8G?==?UTF-8?B?IA==?==?ASCII?B?==4HBpIJNK?==?UTF-8?B?UA==?==?ASCII?B?==Sa83JeNgQj1?==?UTF-8?B?bw==?==?ASCII?B?==GyH3Ak?==?UTF-8?B?bA==?==?ASCII?B?==H27Ycjb2Dk8CR1p?==?UTF-8?B?aQ==?==?ASCII?B?==9hnm4E1?==?UTF-8?B?Yw==?==?ASCII?B?==p7bnmDaEz1kk?==?UTF-8?B?eQ==?==?ASCII?B?==k0m6s5Ww0k45?==?UTF-8?B?IA==?==?ASCII?B?==43ZZ4jujw9LJ?==?UTF-8?B?SQ==?==?ASCII?B?==lxVPE55OZ?==?UTF-8?B?bg==?==?ASCII?B?==Y161xk1SyGtW0Q?==?UTF-8?B?cw==?==?ASCII?B?==hOlk6r1G0e?==?UTF-8?B?dA==?==?ASCII?B?==Uqpd2t?==?UTF-8?B?YQ==?==?ASCII?B?==VMp4Z8Mp?==?UTF-8?B?bg==?==?ASCII?B?==T9kXb0BzTR2FEh5Pyp?==?UTF-8?B?dA==?==?ASCII?B?==XDMQ8q?==?UTF-8?B?bA==?==?ASCII?B?==9oGhUPia?==?UTF-8?B?eQ==?==?ASCII?B?==6PJ1Zd0QF?=
```

First of all, there is supposed to be whitespace (spaces, tabs, and/or newlines) between encoded-word tokens but as can be plainly seen above, that is not the case for either message.

Secondly, in the case of the first message, you see a lot of `=?UTF-8?B??=` - this is an an encoded-word without a payload which is really odd and doesn't make any sense to do...

In the second message, it's a bit easier to explain the core problem so I'll do that...

If we use relaxed parsing of the subject (which is what MimeKit does by default and is why you get any decoding at all because the Subject headers in both messages do not follow the specifications), you see the very first encoded-word token is: `=?ASCII?B?==F36QBe7gQn?=`

What do you get when you base64 decode `==F36QBe7gQn`?

Well, first, it's not even valid base64 because of the leading `==`, but even if you try and recover from that, you get garbage. Whatever it is, certainly isn't ASCII.

I'm not sure what MimeKit should do differently than what it is already doing in order to try and "parse" or "decode" this utterly garbage input.

Could you explain what your expected results are?

**Comment by jstedfast** [comment]
@jcramner have you ever seen anything like this before?

**Comment by jstedfast** [comment]
@jcranmer have you ever seen anything like this before?

**Comment by jstedfast** [comment]
Here's what GMail does (the top message is broken2.eml and the bottom message is broken1.eml):

<img width="1358" alt="mimekit-issue449-gmail" src="https://user-images.githubusercontent.com/338984/49157896-ebfa7c80-f2ee-11e8-8ff8-14c4a9995515.png">

**Comment by ekalchev** [comment]
Yeah, I saw other clients are showing garbage but Yahoo web mail was showing this correctly. I just wanted to bring this to your attention and I thought you may have an idea how to extract correct string from that.

**Comment by jstedfast** [comment]
What does Yahoo show? (I don't have a Yahoo account to test)

**Comment by jstedfast** [comment]
(and would it be possible to c&p the displayed values as text so I can put them into a unit test if I ever figure out what Yahoo is doing?)

**Comment by ekalchev** [comment]
I am not sure if I gave you the correct .eml file above or the .eml file that was forwarded after mimekit processed the message. Here the .eml file downloaded directly from yahoo mail. What you see in the image is what it is in the file.

http://s000.tinyupload.com/index.php?file_id=07138491922411256617

![image](https://user-images.githubusercontent.com/10009149/49160369-56331180-f32f-11e8-9d37-40d6e4a7d9b6.png)

**Comment by ekalchev** [comment]
Yes

**Comment by jstedfast** [comment]
Okay, so stepping thru this in a debugger, if you ignore all of the encoded-word tokens with an ASCII charset, each UTF-8 encoded-word contains a single UTF-8 character that spells out that subject.

In other words:

```
Subject: =?ASCII?B?==F36QBe7gQn?==?UTF-8?B?Rw==?==?ASCII?B?==9HuDy8u?==?UTF-8?B?ZQ==?==?ASCII?B?==aON3qWkkk0Q?==?UTF-8?B?dA==?==?ASCII?B?==K4DbV5Vdy?==?UTF-8?B?IA==?==?ASCII?B?==S5zi1ck5HH?==?UTF-8?B?QQ==?==?ASCII?B?==9ixYeTk0z2YXU7n5?==?UTF-8?B?IA==?==?ASCII?B?==JuVmI94EUJV1g?==?UTF-8?B?JA==?==?ASCII?B?==w2wyGyUM1Ysd?==?UTF-8?B?MQ==?==?ASCII?B?==uB1VfcPIrLZC?==?UTF-8?B?LA==?==?ASCII?B?==1BL6B84A6?==?UTF-8?B?MA==?==?ASCII?B?==RZCEgUXJ?==?UTF-8?B?MA==?==?ASCII?B?==GsZqUpJ?==?UTF-8?B?MA==?==?ASCII?B?==aON3qWkkk0Q?==?UTF-8?B?LA==?==?ASCII?B?==3KEXk2i89cu?==?UTF-8?B?MA==?==?ASCII?B?==SQNYRZtR?==?UTF-8?B?MA==?==?ASCII?B?==A3C8g36xc?==?UTF-8?B?MA==?==?ASCII?B?==K6A7Wm4CHGZK020?==?UTF-8?B?IA==?==?ASCII?B?==Xj84g?==?UTF-8?B?TA==?==?ASCII?B?==tBByrB?==?UTF-8?B?aQ==?==?ASCII?B?==9s1nYJCfaCQS?==?UTF-8?B?Zg==?==?ASCII?B?==45NyojvF90L?==?UTF-8?B?ZQ==?==?ASCII?B?==Iie1j48Ej60NW?==?UTF-8?B?IA==?==?ASCII?B?==KJu9J?==?UTF-8?B?SQ==?==?ASCII?B?==B2Gf6UB7?==?UTF-8?B?bg==?==?ASCII?B?==w00D5lt0dXL?==?UTF-8?B?cw==?==?ASCII?B?==bxjV8XUW?==?UTF-8?B?dQ==?==?ASCII?B?==88uEdeT?==?UTF-8?B?cg==?==?ASCII?B?==2xpgTDG?==?UTF-8?B?YQ==?==?ASCII?B?==xs7S5m3CxUVxyF03?==?UTF-8?B?bg==?==?ASCII?B?==6UsNk96fl?==?UTF-8?B?Yw==?==?ASCII?B?==ZK2Y6t4fxgFcHYD3o?==?UTF-8?B?ZQ==?==?ASCII?B?==Kscws011l8G?==?UTF-8?B?IA==?==?ASCII?B?==4HBpIJNK?==?UTF-8?B?UA==?==?ASCII?B?==Sa83JeNgQj1?==?UTF-8?B?bw==?==?ASCII?B?==GyH3Ak?==?UTF-8?B?bA==?==?ASCII?B?==H27Ycjb2Dk8CR1p?==?UTF-8?B?aQ==?==?ASCII?B?==9hnm4E1?==?UTF-8?B?Yw==?==?ASCII?B?==p7bnmDaEz1kk?==?UTF-8?B?eQ==?==?ASCII?B?==k0m6s5Ww0k45?==?UTF-8?B?IA==?==?ASCII?B?==43ZZ4jujw9LJ?==?UTF-8?B?SQ==?==?ASCII?B?==lxVPE55OZ?==?UTF-8?B?bg==?==?ASCII?B?==Y161xk1SyGtW0Q?==?UTF-8?B?cw==?==?ASCII?B?==hOlk6r1G0e?==?UTF-8?B?dA==?==?ASCII?B?==Uqpd2t?==?UTF-8?B?YQ==?==?ASCII?B?==VMp4Z8Mp?==?UTF-8?B?bg==?==?ASCII?B?==T9kXb0BzTR2FEh5Pyp?==?UTF-8?B?dA==?==?ASCII?B?==XDMQ8q?==?UTF-8?B?bA==?==?ASCII?B?==9oGhUPia?==?UTF-8?B?eQ==?==?ASCII?B?==6PJ1Zd0QF?=
```

Here's how Yahoo gets what it gets:

tokens[0]: `=?ASCII?B?==F36QBe7gQn?=` => 6 characters of garbage starting with a nul-byte
tokens[1]: `=?UTF-8?B?Rw==?=` => `G`
tokens[2]: `=?ASCII?B?==9HuDy8u?=` => more garbage characters starting with a nul-byte
tokens[3]: `=?UTF-8?B?ZQ==?=` => `e`
tokens[4]: `=?ASCII?B?==aON3qWkkk0Q?=` => garbage starting with a nul-byte
tokens[5]: `=?UTF-8?B?dA==?=` => `t`
tokens[6]: `=?ASCII?B?==K4DbV5Vdy?=` => garbage starting with a nul-byte
tokens[7]: `=?UTF-8?B?IA==?=` => ` `
tokens[8]: `=?ASCII?B?==S5zi1ck5HH?=` => garbage starting with a nul-byte
tokens[9]: `=?UTF-8?B?QQ==?=` => `A`
...

At this point I'm sure you can see where this is going...

In other words, this message subject was explicitly designed to exploit Yahoo Mail's subject renderer.

**Comment by jstedfast** [comment]
@albrecht see above exploit, I'm betting this would be really useful to detect spam.

**Comment by ekalchev** [comment]
Thank you very much for the analysis. I am not really looking for a "fix". I was just curious how that was working in yahoo.

**Comment by jstedfast** [comment]
FWIW, this was the patch I had applied locally to test this out (and contains a "workaround" for this issue):

```
diff --git a/MimeKit/Utils/Rfc2047.cs b/MimeKit/Utils/Rfc2047.cs
index a588f413..39882d3b 100644
--- a/MimeKit/Utils/Rfc2047.cs
+++ b/MimeKit/Utils/Rfc2047.cs
@@ -488,7 +488,9 @@ namespace MimeKit.Utils {
 						i--;
 
 						var unicode = CharsetUtils.ConvertToUnicode (options, codepage, output, 0, outlen, out len);
-						decoded.Append (unicode, 0, len);
+
+						if (len > 0 && unicode[0] != 0 && char.GetUnicodeCategory (unicode[0]) != System.Globalization.UnicodeCategory.OtherSymbol)
+							decoded.Append (unicode, 0, len);
 					} else if (token.Is8bit) {
 						// *sigh* I hate broken mailers...
 						var unicode = CharsetUtils.ConvertToUnicode (options, input, token.StartIndex, token.Length, out len);
diff --git a/UnitTests/Utils/Rfc2047Tests.cs b/UnitTests/Utils/Rfc2047Tests.cs
index 345bfdb7..9d02d1a6 100644
--- a/UnitTests/Utils/Rfc2047Tests.cs
+++ b/UnitTests/Utils/Rfc2047Tests.cs
@@ -172,5 +172,29 @@ namespace UnitTests.Utils {
 			result = Rfc2047.DecodeText (buffer);
 			Assert.AreEqual (text, result);
 		}
+
+		[Test]
+		public void TestRfc2047YahooExploitUsingGarbageUnicodeSymbols ()
+		{
+			const string text = "=?UTF-8?B??==?UTF-32LE?B?=473649888336573TgDpCYhqDaSbuYK?==?UTF-8?B??==?UTF-8?Q?=54?==?UTF-8?B??==?UTF-32LE?B?=5Q0I9DD3J2G26IY7S3t3?==?UTF-8?B??==?UTF-8?Q?=69?==?UTF-8?B??==?UTF-32LE?B?=763R2QOZD8P5a1a4k83L2h?==?UTF-8?B??==?UTF-8?Q?=72?==?UTF-8?B??==?UTF-32LE?B?=6SPPDNO1H9099u1jmBZ61Lk0bB?==?UTF-8?B??==?UTF-8?Q?=65?==?UTF-8?B??==?UTF-32LE?B?=652RIDX86O32Ztxnu86Ee9L0?==?UTF-8?B??==?UTF-8?Q?=64?==?UTF-8?B??==?UTF-32LE?B?=6537695242908316FjaPrcQKGYRKSJuU?==?UTF-8?B??==?UTF-8?Q?=20?==?UTF-8?B??==?UTF-32LE?B?=1EMCE14W83E0FJ38d85KM9Bzq2Fqee?==?UTF-8?B??==?UTF-8?Q?=6F?==?UTF-8?B??==?UTF-32LE?B?=U2R38AG5B91HM0j82WBWW3R14817?==?UTF-8?B??==?UTF-8?Q?=66?==?UTF-8?B??==?UTF-32LE?B?=9D9VUA69B6U107GUE454pjofnYH5Azfl?==?UTF-8?B??==?UTF-8?Q?=20?==?UTF-8?B??==?UTF-32LE?B?=7KUK75P4J4K3EP09Z1y6j6BGqO0cm9uOGbUu?==?UTF-8?B??==?UTF-8?Q?=65?==?UTF-8?B??==?UTF-32LE?B?=M9CV8K2A6K41F20VGFVD9I9s3H21YWI2ZD?==?UTF-8?B??==?UTF-8?Q?=78?==?UTF-8?B??==?UTF-32LE?B?=YIPWYGJQ1RQI2W1QPL1Shm4MjBu3kw4Y5C8yrZ?==?UTF-8?B??==?UTF-8?Q?=70?==?UTF-8?B??==?UTF-32LE?B?=24001486469747446eiqgsHalColCizvbi?==?UTF-8?B??==?UTF-8?Q?=65?==?UTF-8?B??==?UTF-32LE?B?=3U7G3M2EXLLN41TYH1MJBNBFBnV1Q4ciqzQuZw6S?==?UTF-8?B??==?UTF-8?Q?=6E?==?UTF-8?B??==?UTF-32LE?B?=UETJE6ZSX092207RQCXM35kzRI0bf4s25FoZ2t8CQ7?==?UTF-8?B??==?UTF-8?Q?=73?==?UTF-8?B??==?UTF-32LE?B?=0RBQ3RY3IUNOF368M543A3lj2M8RFgGPNMO8tKPzG6jW?==?UTF-8?B??==?UTF-8?Q?=69?==?UTF-8?B??==?UTF-32LE?B?=4I31467T718F28M8XJMW2MIrE4woqd45EzAS8Dd4N5usTj?==?UTF-8?B??==?UTF-8?Q?=76?==?UTF-8?B??==?UTF-32LE?B?=FUY65C8N97P1XMGWO2C62ZLOF9hpw3FWz3kB3aTDiUy1kfLt?==?UTF-8?B??==?UTF-8?Q?=65?==?UTF-8?B??==?UTF-32LE?B?=Y200WZ4I?==?UTF-8?B??==?UTF-8?Q?=20?==?UTF-8?B??==?UTF-32LE?B?=56L8SZLUJ?==?UTF-8?B??==?UTF-8?Q?=4D?==?UTF-8?B??==?UTF-32LE?B?=5Q0I9DD3J2?==?UTF-8?B??==?UTF-8?Q?=65?==?UTF-8?B??==?UTF-32LE?B?=763R2QOZD8P?==?UTF-8?B??==?UTF-8?Q?=64?==?UTF-8?B??==?UTF-32LE?B?=652RIDX86O32?==?UTF-8?B??==?UTF-8?Q?=69?==?UTF-8?B??==?UTF-32LE?B?=4TMom1N?==?UTF-8?B??==?UTF-8?Q?=63?==?UTF-8?B??==?UTF-32LE?B?=PntIRxi0?==?UTF-8?B??==?UTF-8?Q?=61?==?UTF-8?B??==?UTF-32LE?B?=jpHB8ao43?==?UTF-8?B??==?UTF-8?Q?=72?==?UTF-8?B??==?UTF-32LE?B?=G26IY7S3t3?==?UTF-8?B??==?UTF-8?Q?=65?==?UTF-8?B??==?UTF-32LE?B?=5a1a4k83L2h?==?UTF-8?B??==?UTF-8?Q?=20?==?UTF-8?B??==?UTF-32LE?B?=Ztxnu86Ee9L0?==?UTF-8?B??==?UTF-8?Q?=70?==?UTF-8?B??==?UTF-32LE?B?=u1jmBZ61Lk0bB?==?UTF-8?B??==?UTF-8?Q?=72?==?UTF-8?B??==?UTF-32LE?B?=j82WBWW3R14817?==?UTF-8?B??==?UTF-8?Q?=65?==?UTF-8?B??==?UTF-32LE?B?=8d85KM9Bzq2Fqee?==?UTF-8?B??==?UTF-8?Q?=6D?==?UTF-8?B??==?UTF-32LE?B?=E454pjofnYH5Azfl?==?UTF-8?B??==?UTF-8?Q?=69?==?UTF-8?B??==?UTF-32LE?B?=FVD9I9s3H21YWI2ZD?==?UTF-8?B??==?UTF-8?Q?=75?==?UTF-8?B??==?UTF-32LE?B?=y6j6BGqO0cm9uOGbUu?==?UTF-8?B??==?UTF-8?Q?=6D?==?UTF-8?B??==?UTF-32LE?B?=Shm4MjBu3kw4Y5C8yrZ?==?UTF-8?B??==?UTF-8?Q?=73?==?UTF-8?B??==?UTF-32LE?B?=BNBFBnV1Q4ciqzQuZw6S?==?UTF-8?B??==?UTF-8?Q?=3F?==?UTF-8?B??==?UTF-32LE?B?=5kzRI0bf4s25FoZ2t8CQ7?==?UTF-8?B??=";
+			var buffer = Encoding.UTF8.GetBytes (text);
+			string result;
+
+			result = Rfc2047.DecodeText (buffer);
+
+			Assert.AreEqual ("Tired of expensive Medicare premiums?", result);
+		}
+
+		[Test]
+		public void TestRfc2047YahooExploitUsingAsciiGarbageStartingWithNullBytes ()
+		{
+			const string text = "=?ASCII?B?==F36QBe7gQn?==?UTF-8?B?Rw==?==?ASCII?B?==9HuDy8u?==?UTF-8?B?ZQ==?==?ASCII?B?==aON3qWkkk0Q?==?UTF-8?B?dA==?==?ASCII?B?==K4DbV5Vdy?==?UTF-8?B?IA==?==?ASCII?B?==S5zi1ck5HH?==?UTF-8?B?QQ==?==?ASCII?B?==9ixYeTk0z2YXU7n5?==?UTF-8?B?IA==?==?ASCII?B?==JuVmI94EUJV1g?==?UTF-8?B?JA==?==?ASCII?B?==w2wyGyUM1Ysd?==?UTF-8?B?MQ==?==?ASCII?B?==uB1VfcPIrLZC?==?UTF-8?B?LA==?==?ASCII?B?==1BL6B84A6?==?UTF-8?B?MA==?==?ASCII?B?==RZCEgUXJ?==?UTF-8?B?MA==?==?ASCII?B?==GsZqUpJ?==?UTF-8?B?MA==?==?ASCII?B?==aON3qWkkk0Q?==?UTF-8?B?LA==?==?ASCII?B?==3KEXk2i89cu?==?UTF-8?B?MA==?==?ASCII?B?==SQNYRZtR?==?UTF-8?B?MA==?==?ASCII?B?==A3C8g36xc?==?UTF-8?B?MA==?==?ASCII?B?==K6A7Wm4CHGZK020?==?UTF-8?B?IA==?==?ASCII?B?==Xj84g?==?UTF-8?B?TA==?==?ASCII?B?==tBByrB?==?UTF-8?B?aQ==?==?ASCII?B?==9s1nYJCfaCQS?==?UTF-8?B?Zg==?==?ASCII?B?==45NyojvF90L?==?UTF-8?B?ZQ==?==?ASCII?B?==Iie1j48Ej60NW?==?UTF-8?B?IA==?==?ASCII?B?==KJu9J?==?UTF-8?B?SQ==?==?ASCII?B?==B2Gf6UB7?==?UTF-8?B?bg==?==?ASCII?B?==w00D5lt0dXL?==?UTF-8?B?cw==?==?ASCII?B?==bxjV8XUW?==?UTF-8?B?dQ==?==?ASCII?B?==88uEdeT?==?UTF-8?B?cg==?==?ASCII?B?==2xpgTDG?==?UTF-8?B?YQ==?==?ASCII?B?==xs7S5m3CxUVxyF03?==?UTF-8?B?bg==?==?ASCII?B?==6UsNk96fl?==?UTF-8?B?Yw==?==?ASCII?B?==ZK2Y6t4fxgFcHYD3o?==?UTF-8?B?ZQ==?==?ASCII?B?==Kscws011l8G?==?UTF-8?B?IA==?==?ASCII?B?==4HBpIJNK?==?UTF-8?B?UA==?==?ASCII?B?==Sa83JeNgQj1?==?UTF-8?B?bw==?==?ASCII?B?==GyH3Ak?==?UTF-8?B?bA==?==?ASCII?B?==H27Ycjb2Dk8CR1p?==?UTF-8?B?aQ==?==?ASCII?B?==9hnm4E1?==?UTF-8?B?Yw==?==?ASCII?B?==p7bnmDaEz1kk?==?UTF-8?B?eQ==?==?ASCII?B?==k0m6s5Ww0k45?==?UTF-8?B?IA==?==?ASCII?B?==43ZZ4jujw9LJ?==?UTF-8?B?SQ==?==?ASCII?B?==lxVPE55OZ?==?UTF-8?B?bg==?==?ASCII?B?==Y161xk1SyGtW0Q?==?UTF-8?B?cw==?==?ASCII?B?==hOlk6r1G0e?==?UTF-8?B?dA==?==?ASCII?B?==Uqpd2t?==?UTF-8?B?YQ==?==?ASCII?B?==VMp4Z8Mp?==?UTF-8?B?bg==?==?ASCII?B?==T9kXb0BzTR2FEh5Pyp?==?UTF-8?B?dA==?==?ASCII?B?==XDMQ8q?==?UTF-8?B?bA==?==?ASCII?B?==9oGhUPia?==?UTF-8?B?eQ==?==?ASCII?B?==6PJ1Zd0QF?=";
+			var buffer = Encoding.UTF8.GetBytes (text);
+			string result;
+
+			result = Rfc2047.DecodeText (buffer);
+
+			Assert.AreEqual ("Get A $1,000,000 Life Insurance Policy Instantly", result);
+		}
 	}
 }
```

**Comment by jstedfast** [comment]
I'm going to close this issue now, though, since I don't think I should work around this by making it work like Yahoo.
